### PR TITLE
Move boost asio socket interface to socketif_t

### PR DESCRIPTION
**Pull request by jerryz123**
This PR reduces the complexity of linking against spike, addressing some of the issues discussed in #1168.

All boost-asio objects are moved to a `socketif_t` member of `sim_t`, moving spike towards a world where the interface of `sim_t` does not depend on `configure`-time settings.

### CODE REVIEW

File: riscv/socketif.cc [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      io_service_ptr = new boost::asio::io_service;
+      acceptor_ptr = new tcp::acceptor(*io_service_ptr, tcp::endpoint(tcp::v4(), 0));
+      // acceptor is created passing argument port=0, so O.S. will choose a free port
+      std::string name = boost::asio::ip::host_name();
+      std::cout << "Listening for debug commands on " << name.substr(0,name.find('.'))
+                << " port " << acceptor_ptr->local_endpoint().port() << " ." << std::endl;
+      // at the end, add space and some other character for convenience of javascript .split(" ")
+    }
+  catch (std::exception& e)
+    {
```
Comment (aswaterman) [inline riscv/socketif.cc]: Looks like the brace formatting in these blocks got screwed up.  Should change it to match the subsequent blocks in this file, which conform to our house style.

File: riscv/sim.cc [hunk 2]
```diff
@@ -74,9 +70,23 @@ sim_t::sim_t(const cfg_t *cfg, bool halted,
...
+         "please rebuild the riscv-isa-sim project without --without-boost-asio\n",
+         stderr);
+    abort();
+  }
+#endif
+
 #ifndef RISCV_ENABLE_DUAL_ENDIAN
   if (cfg->endianness != memif_endianness_little) {
-    fputs("Big-endian support has not been prroperly enabled; "
+    fputs("Big-endian support has not been properly enabled; "
```
Comment (aswaterman) [inline riscv/sim.cc]: This sort of change should be in its own commit; refactoring commits should not have functional changes (even trivial ones) when possible.

**Review by aswaterman — COMMENTED** [review]
This certainly seems like a step in the right direction.  I'll give @scottj97 a chance to review and merge if he feels it's appropriate.

**Comment by aswaterman** [comment]
BTW, the CI flow only tests without Boost.  I assume you tested with Boost, as well?

**Comment by jerryz123** [comment]
> BTW, the CI flow only tests without Boost. I assume you tested with Boost, as well?

I ran the ci-tests script on my machine, which does have boost asio. I also stepped through a few instructions in `-d` mode with boost enabled, and confirmed that `spike -s` does go to `Listening for debug commands on ...`. I didn't actually test connecting to the socket though, I don't know what the socket client is supposed to be.

**Comment by aswaterman** [comment]
Cool.

File: riscv/sim.cc [hunk 3]
```diff
@@ -74,6 +70,20 @@ sim_t::sim_t(const cfg_t *cfg, bool halted,
...
 
+  socketif = NULL;
+#ifdef HAVE_BOOST_ASIO
+  if (socket_enabled) {
+    socketif = new socketif_t();
+  }
+#else
+  if (socket_enabled) {
+    fputs("Socket support requires compilation with boost asio; "
+          "please rebuild the riscv-isa-sim project without --without-boost-asio\n",
```
Comment (scottj97) [inline riscv/sim.cc]: This could be confusing to a user who doesn't have Boost installed, and did not use that option to `configure`.

Would it be correct to say `rebuild with --with-boost-asio`? Then `configure` will tell them that the Boost package is missing. (Right?)

**Review by scottj97 — CHANGES_REQUESTED** [review]

**Review by scottj97 — APPROVED** [review]

**Review by aswaterman — APPROVED** [review]

**Comment by jerryz123** [comment]
Rebased, should be good to go now.
